std/typecons.d

```diff
@@ -2440,33 +2442,65 @@ Params:
         assert(e != 12);
     }
 
-    template toString()
+    /**
+     * Gives the string `"Nullable.null"` if `isNull` is `true`. Otherwise, the
+     * result is equivalent to calling $(REF formattedWrite, std,format) on the
+     * underlying value.
+     *
+     * Params:
+     *     writer = A `char` accepting
+     *     $(REF_ALTTEXT output range, isOutputRange, std, range, primitives)
+     *     fmt = A $(REF FormatSpec, std,format) which is used to represent
+     *     the value if this Nullable is not null
+     * Returns:
+     *     A `string` if `writer` and `fmt` are not set; `void` otherwise.
+     */
+    string toString()()
```

Comment:
This can at least be made `pure` and `@safe`. Is there a reason you made this overload a template instead of a regular function and marking it with those attributes? Inference doesn't really buy us anything in this case because there are no template arguments that the attributes could depend on.

Reply:
The `toString` of `_value` can be anything.

Reply:
@MetaLang It can depend on template parameters on the containing struct.

It's actually a bad idea to put explicit attributes on templates, and members of templated aggregates, because that constrains what types the template can be instantiated with.  It should be rather left up to the compiler to infer them, and unittests to enforce mandatory attributes for specific instantiations of the template.

Reply:
Fair enough, but I don't think making it a no-arg template is necessary as it's inside a templated struct.
```
struct Test()
{
    string toString()
    {
        auto app = appender!string();
        auto spec = singleSpec("%s");
        toString(app, spec);
        return app.data;
    }

    /// ditto
    void toString(W)(ref W writer, const ref FormatSpec!char fmt)
        if (isOutputRange!(W, char))
    {
        import std.range.primitives : put;
        put(writer, "Nullable.null");
    }
}

void main()
{
	import std.stdio;
    import std.traits;
    alias FA = FunctionAttribute;
   	writeln(hasFunctionAttributes!(Test!().toString, "@safe", "pure")); //Prints "true"
    writeln(!hasFunctionAttributes!(Test!().toString, "nothrow", "@nogc", "@system")); //Prints "false"
}
```
As you can see, the inference is done even though `toString` is not an explicitly templated function. Put in a call to an `@system` function and this will print "false" and "true" instead.